### ISSUE TITLE
After test run, output is not generated at location evosuite-report/

### ISSUE THREAD

**Issue by AaratiT** [EvoSuite/evosuite#213]
I am trying to follow the evosuite tutorial (http://www.evosuite.org/documentation/tutorial-part-1/)

Am able to generate test class, compile & run it.
However the report generated after test run is not getting updated in evosuite-report/statistics.csv
(Please note, when test class is generated, the CSV file is created & has few entries in 'statistics' with columns like 
TARGET_CLASS | criterion | Coverage | Total_Goals | Covered_Goals
-- | -- | -- | -- | --
)
I am  [referring ](http://www.evosuite.org/documentation/commandline/)  to know what kind of EvoSuite Output is generated after test run.

### Steps to Reproduce
1. Using java8,  use -Dsandbox_mode=OFF option to generate the test classes
2.  Command to run test classes
java  -cp ".\log4j.jar;.\ojdbc6.jar;.\commons-codec-1.3.jar;.\slf4j-log4j12-1.7.16.jar;.\slf4j-api-1.7.16.jar;D:\g\k\java;D:\proj\target\classes;.\evosuite-tests;.\evosuite-standalone-runtime-1.0.6.jar;.\junit-4.12.jar;.\hamcrest-core-1.3.jar;"   org.junit.runner.JUnitCore  com.u.g.KeyClient_ESTest
3. Output
JUnit version 4.12
.log4j:ERROR Failed to flush writer,
org.evosuite.runtime.mock.java.io.NativeMockedIO.writeBytes(NativeMockedIO.java:78)
<evosuite>.<evosuite>(<evosuite>)
<evosuite>.<evosuite>(<evosuite>)
log4j:ERROR Failed to flush writer,
org.evosuite.runtime.mock.java.io.NativeMockedIO.writeBytes(NativeMockedIO.java:78)
<evosuite>.<evosuite>(<evosuite>)
<evosuite>.<evosuite>(<evosuite>)
.log4j:ERROR Failed to flush writer,
org.evosuite.runtime.mock.java.io.NativeMockedIO.writeBytes(NativeMockedIO.java:78)
<evosuite>.<evosuite>(<evosuite>)
<evosuite>.<evosuite>(<evosuite>)
.log4j:ERROR Failed to flush writer,
org.evosuite.runtime.mock.java.io.NativeMockedIO.writeBytes(NativeMockedIO.java:78)
<evosuite>.<evosuite>(<evosuite>)
<evosuite>.<evosuite>(<evosuite>)

Time: 4.648

OK (3 tests)

### Expected result
EvoSuite Output  at evosuite-report/data is not generated after test run.

Many thanks.

**Comment by jose** [comment]
Hi @AaratiT,

First, why are you using `-Dsandbox_mode=OFF` ?

Second, the execution of test cases generated by EvoSuite does not update the evosuite-report/statistics.csv. The `statistics.csv` file you see under `evosuite-report/` is **only** generated/updated when you run EvoSuite, and not when you run test cases generated by EvoSuite.

--
Cheers,
Jose

**Comment by AaratiT** [comment]
Hi Jose,

Thanks for writing back.
1. Using sandbox off option because when I was running the test case of a 
class which was making a JDBC call to DB, I was getting below exception

java.lang.SecurityException: Security manager blocks 
("java.sql.SQLPermission" "deregisterDriver")

2. Now after the test run is completed, how do we know the outcome i,e any 
kind of report or anything which indicates test execution summary report.
As per below , I though there will be some way to know that  (
http://www.evosuite.org/documentation/commandline/)
A report consists of CSV data files (evosuite-report/statistics.csv 
contains the data collected after each run, and evosuite-report/data 
ontains one CSV file for each run with detailed information), and 
optionally an HTML report (evosuite-report/report_generation.html). The 
HTML report contains one entry for each run (i.e. one for each run on a 
class), listing all the parameters, the test cases, the coverage, and if 
plot=true in evosuite.properties it also contains plots of the fitness, 
length, and size histories. 

So can you please suggest command line options I should try to get the 
report.
Many Thanks.

Best,
Aarati

**Comment by jose** [comment]
Hi @AaratiT,

> 2. Now after the test run is completed, how do we know the outcome i,e any
> kind of report or anything which indicates test execution summary report.
> ...

Once again, the section in the command line tutorial you're quoting describes the output produced by EvoSuite, and not the output produced by the generated test cases when they are executed on the command line!

What kind of execution summary are you looking for? Result of each test case? Runtime of each test case? Coverage of the generated test suite? Mutation score of the generated test suite?

As you're running test cases on the command line, there aren't much information you can extract... However, if you switch to [Maven](https://maven.apache.org), there are several plugins you can use to collect data of the generated test suite. For instance, [maven-surefire-plugin](https://maven.apache.org/surefire/maven-surefire-plugin/) generates plain text reports with the result and runtime of each test case; if you want to measure the coverage of the generated test cases you may want to integrate a tool called JaCoCo (more info in [here](http://www.evosuite.org/documentation/measuring-code-coverage/)), etc.


--
Best,
Jose

**Comment by AaratiT** [comment]
Hi Jose,
From command line execution, is there a way to know Result of each test case run? 
Any text, csv file which says TestClassA had 3 test cases, 2 passed and 1 failed.

Also if I modify generated test class to put some assertions like assertNotNull, how do I know if they have passed or failed?
Thanks again.

**Comment by jose** [comment]
> From command line execution, is there a way to know Result of each test case run?
> Any text, csv file which says TestClassA had 3 test cases, 2 passed and 1 failed.

There's a way to get that information but it's a bit tricky, as you would need to parse (e.g., grep) the output produced by the command.

The following command
```
$ java -cp ... org.junit.runner.JUnitCore com.u.g.KeyClient_ESTest
```
prints out
```
...
OK (X tests)
```
if there aren't any failing test case. (X is the total number of test cases in the KeyClient_ESTest suite).

However, in case there are failing test cases, the above command prints out
```
...
Tests run: X,  Failures: Y
```
(where X is the total number of test cases in the test suite, and Y the number of test cases that failed).

> Also if I modify generated test class to put some assertions like assertNotNull, how do I know if they have passed or failed?

This would be even more difficult to get. For each failing test case the `org.junit.runner.JUnitCore` prints out the stack trace, for example,
```
...
There was 1 failure:
1) test0(org.Calculator_ESTest)
org.junit.ComparisonFailure: expected:<[asdadsa]> but was:<[]>
	at org.junit.Assert.assertEquals(Assert.java:115)
	at org.junit.Assert.assertEquals(Assert.java:144)
	at org.Calculator_ESTest.test0(Calculator_ESTest.java:22)
...
```
you will need to search in the stack trace whether the number of the line with the `assertNotNull` is there or not.

**Comment by AaratiT** [comment]
Hmm.
Have few follow up questions :)
1.  regarding the way   test execution happens ,  i.e  one of the test method calls the actual class method.
Now this class method is suppose to make a JDBC connection with DB & fetches some value. 
So when this test executed, I wanted to know whether actual JDBC connection is created or evosuite mocks it ? 
2. regarding maven approach. In order to view " The HTML report contains one entry for each run ", can you please specific if I follow [this](http://www.evosuite.org/documentation/tutorial-part-2/  ) tutorial, it will give test run report ? or is there any additional settings , parameters required?
Thanks

**Comment by jose** [comment]
> regarding the way test execution happens , i.e one of the test method calls the actual class method.
> Now this class method is suppose to make a JDBC connection with DB & fetches some value.
> So when this test executed, I wanted to know whether actual JDBC connection is created or evosuite mocks it ?

It's very difficult to answer your question without seeing the test case and the class under test...

> regarding maven approach. In order to view " The HTML report contains one entry for each run ", can you please specific if I follow this tutorial, it will give test run report ? or is there any additional settings , parameters required?

As the generation of the `evosuite-report/statistics.csv`, a HTML report is generated by EvoSuite and not by executing the test cases generated by EvoSuite. Nevertheless, if you want EvoSuite to generate a HTML report rather than the `statistics.csv`, you have to use the following option `-Dstatistics_backend=HTML` or `-DextraArgs="-Dstatistics_backend=HTML"` (if you use the EvoSuite maven plugin).

**Comment by AaratiT** [comment]
Here is  the test case and the class under test...

     @RunWith(EvoRunner.class) @EvoRunnerParameters(mockJVMNonDeterminism = true, useVFS = true, useVNET = true, resetStaticState = true, separateClassLoader = true, useJEE = true) 
     public class KeyClient_ESTest extends KeyClient_ESTest_scaffolding {

    @Test(timeout = 4000)
    public void test0()  throws Throwable  {
      // Undeclared exception!
      try { 
        KeyClient.testKey();
        fail("Expecting exception: RuntimeException");
      
      } catch(RuntimeException e) {
         //
         // Could not load resource bundle GDPR.properties. Throwing runtime exception.
         //
         verifyException("com.ultimatix.gdpr.KeyClient", e);
      }
      }
       // there  are 2 more methods
      }
//  class KeyClient has below method. It usese sigleton pattern

               public final class KeyClient {
  
               private static String temp=null;
               public static String testKey() {
		
		ResultSet rs = null;
		PreparedStatement cstmt = null;
		Connection connection = null;
		try {
			
			Class.forName("oracle.jdbc.driver.OracleDriver");															
			String decrpytUser = "user";
			String decrpytPass = "password#";		

			connection = DriverManager.getConnection("jdbcurl", decrpytUser, decrpytPass);
			String query = "select * from table1";
			    cstmt = connection.prepareStatement(query);				
				rs = cstmt.executeQuery();
				while (rs.next()) {
					String s = rs.getString(1);

					if (s != null) {
						String[] as = s.split("\\|@\\|");
						if (as != null && as.length == 3) {
							if (("Failure").endsWith(as[0])) {
								LOG.info("Exception" + as[1]);
							} else {
								LOG.info("Success message" + as[1]);
								temp = as[2];
							}
						}

					}
				}
			

		} catch (SQLException se) {

			throw new RuntimeException("SQL Error! Check output console", se);

		} catch (Exception e) {

			throw new RuntimeException("Some  Error! Check output console", e);

		} finally {
			try {
				if (null != rs && !rs.isClosed()) {
					rs.close();
				}
				if (null != cstmt && !cstmt.isClosed()) {
					cstmt.close();
				}
				if (null != connection && !connection.isClosed()) {
					connection.close();
				}
			} catch (SQLException e) {
				LOG.error("Connection closure error ! Check output console", e);
			}
		}
		return temp;
	}
        }

From your repose, it appears that test run / execution report is not obtained.
Even in log files, I could not find assertions as stated by you. 
So in short we are really stuck on tow issues. whether  test execution is really meaningful i.e its mocking or actually executing the underline code  & secondly  how to verify assertions.
Any  help in this regards will be highly appreciated.

**Comment by jose** [comment]
As far I can see no object seems to be mocked.

> Even in log files, I could not find assertions as stated by you.

What log files?

PS: I can't compile the class under test, therefore I can't compile/run the test class.

**Comment by AaratiT** [comment]
What log files? the ones used by self4j -  log4j,The log file path is mentioned inside log4j.properties

Posted the class & est class as in  your  earlier comment said 

> It's very difficult to answer your question without seeing the test case and the class under test...

What is the compile  time error u r getting ?

**Comment by jose** [comment]
> What is the compile time error u r getting ?

```
$ javac -cp . KeyClient.java
KeyClient.java:6: error: cannot find symbol
	ResultSet rs = null;
	^
  symbol:   class ResultSet
  location: class KeyClient
KeyClient.java:7: error: cannot find symbol
	PreparedStatement cstmt = null;
	^
  symbol:   class PreparedStatement
  location: class KeyClient
KeyClient.java:8: error: cannot find symbol
	Connection connection = null;
	^
  symbol:   class Connection
  location: class KeyClient
KeyClient.java:15: error: cannot find symbol
		connection = DriverManager.getConnection("jdbcurl", decrpytUser, decrpytPass);
		             ^
  symbol:   variable DriverManager
  location: class KeyClient
KeyClient.java:26: error: cannot find symbol
							LOG.info("Exception" + as[1]);
							^
  symbol:   variable LOG
  location: class KeyClient
KeyClient.java:28: error: cannot find symbol
							LOG.info("Success message" + as[1]);
							^
  symbol:   variable LOG
  location: class KeyClient
KeyClient.java:37: error: cannot find symbol
	} catch (SQLException se) {
	         ^
  symbol:   class SQLException
  location: class KeyClient
KeyClient.java:56: error: cannot find symbol
		} catch (SQLException e) {
		         ^
  symbol:   class SQLException
  location: class KeyClient
KeyClient.java:57: error: cannot find symbol
			LOG.error("Connection closure error ! Check output console", e);
			^
  symbol:   variable LOG
  location: class KeyClient
9 errors
```

**Comment by AaratiT** [comment]
hi Jose,
Can u plz add below jars in class path 
ojdbc6.jar
log4j.jar
Also plz add tghis line inside class
           private static final Logger LOG = Logger.getLogger(KeyClient.class);

thanks.

**Comment by AaratiT** [comment]
hi Jose,
may please know if any progress
Or anything suggestion I should be trying out

**Comment by AaratiT** [comment]
As the main question is addressed by Jose 

> As the generation of the evosuite-report/statistics.csv, a HTML report is generated by EvoSuite and not by executing the test cases generated by EvoSuite. Nevertheless, if you want EvoSuite to generate a HTML report rather than the statistics.csv, you have to use the following option -Dstatistics_backend=HTML or -DextraArgs="-Dstatistics_backend=HTML" (if you use the EvoSuite maven plugin).

So opening a new issue for rest of my questions
